sources/ldap: implement nested group parentship sync

krejcar25 opened the pull request:
<!--
👋 Hi there! Welcome.

Please check the Contributing guidelines: https://docs.goauthentik.io/docs/developer-docs/#how-can-i-contribute

⚠️ IMPORTANT: Make sure you are opening this PR from a FEATURE BRANCH, not from your main branch!
If you opened this PR from your main branch, please close it and create a new feature branch instead.
For more information, see: https://docs.goauthentik.io/developer-docs/contributing/#always-use-feature-branches
-->

## Details

<!--
Explain what this PR changes, what the rationale behind the change is, if any new requirements are introduced or any breaking changes caused by this PR.

Ideally also link an Issue for context that this PR will close using `closes #`
-->
Release 2025.12 brought many-to-many group inheritance. This PR builds on that, and adds true nested group synchronisation from LDAP. `LDAPSource` model fields were renamed to reflect this, and membership sync now runs through both user and group objects in LDAP reflects directory inheritance structure in Authentik.

If accepted, this PR resolves #9460.

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

If applicable

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make docs`)

krejcar25 committed:
sources/ldap: add group parentage synchronizer

krejcar25 committed:
sources/ldap: regenerate schema after adding field

krejcar25 committed:
web/admin: add LDAP source toggle for sync_group_parentage

krejcar25 committed:
sources/ldap: add sync_group_parentage field

krejcar25 committed:
web/admin: update admin UI with new field names

Also added better descriptions to all relevant fields, explaining how they interact with each other.

krejcar25 committed:
sources/ldap: rename LDAPSource fields to more accurate names

krejcar25 committed:
sources/ldap: add unit tests

krejcar25 committed:
sources/ldap: fix problem discovered by test

krejcar25 committed:
sources/ldap: lint

krejcar25 committed:
web/admin: lint

krejcar25 committed:
website/docs: update LDAP source docs about lookup fields & membership

dominic-r commented on website/docs/users-sources/sources/directory-sync/active-directory/index.md:
```diff
@@ -64,13 +64,14 @@ To support the integration of authentik with Active Directory, you will need to
-    - **Group membership field**: the user object attribute or the group object attribute that determines the group membership of a user (e.g. `member`). If **Lookup using a user attribute** is set, this should be a user object attribute, otherwise a group object attribute.
-    - **User membership attribute**: ensure that this is set to `distinguishedName`.
+    - **Lookup using a member attribute**: acquire group membership from a User object attribute (`memberOf`) instead of a Group attribute (`member`).
+    - **Membership field**: ensure that this is set to `member` for Active Directory nested group lookups, unless flattening memberships is needed. Flattening can be achieved my setting this to `memberOf:1.2.840.113556.1.4.1941:` and enabling **Lookup using a member attribute**.
```
```suggestion
    - **Membership field**: ensure that this is set to `member` for Active Directory nested group lookups, unless flattening memberships is needed. Flattening can be achieved by setting this to `memberOf:1.2.840.113556.1.4.1941:` and enabling **Lookup using a member attribute**.
```

BeryJu commented:
hey @krejcar25 since the above blocking PR is merged, are you interested in continuing with this PR? if not are you ok with us taking over and continuing it?

krejcar25 commented:
@BeryJu just rebased onto current main. I'd like to try implementing this if I can.

BeryJu commented:
Sounds good, feel free to tag any of us for questions

krejcar25 committed:
sources/ldap: fix renamed attributes

krejcar25 committed:
web/admin: make web

krejcar25 committed:
sources/ldap: update tests and entries.json fixture

krejcar25 commented:
As it is now the parentship sync is doing what it's supposed to - syncing group parents, and maintaining any parentship relationships where one side is a non-LDAP group (Domain Admins added into ak-admins, for example).
User lookup still needs to be rewritten to use UserLDAPSourceConnection objects instead of dynamic attributes.

krejcar25 committed:
sources/ldap: update migration with current help_text

krejcar25 committed:
sources/ldap: fix lint error

krejcar25 committed:
sources/ldap: restore membership.py to main, with renamed fields

in prep for separating out the synchronizers

krejcar25 committed:
sources/ldap: abstract get_group into BaseMembershipLDAPSynchronizer

krejcar25 committed:
sources/ldap: add separate ParentshipLDAPSynchronizer

krejcar25 committed:
sources/ldap: fix wrong attribute name in extra parents filtering

gergosimonyi committed:
Merge branch 'main' into sync-group-parents-from-ldap

gergosimonyi committed:
Merge branch 'main' into sync-group-parents-from-ldap

gergosimonyi committed:
Merge branch 'main' into sync-group-parents-from-ldap

gergosimonyi committed:
remove renames

gergosimonyi committed:
rename `sync_group_parents` to `sync_group_hierarchy`

gergosimonyi committed:
set `sync_group_hierarchy` default to `False`

This is a breaking change if it's `True`. It should be changed
eventually, but not in this release.

gergosimonyi committed:
revert pagination function change

I'm not entirely sure why, but this change hangs the test
`test_membership_sync_special_chars_in_group_dn`.

gergosimonyi committed:
add `sync_group_hierarchy` guard to hierarchy sync

gergosimonyi committed:
simplify hierarchy sync

gergosimonyi committed:
reword group `parentship` to `hierarchy`

gergosimonyi committed:
fix lint

gergosimonyi commented:
I've removed the naming changes. They make 100% sense, but they should be part of a separate PR (and probably a separate release).

gergosimonyi committed:
remove dead code

gergosimonyi committed:
move `syncGroupHierarchy` toggle next to similar toggles

kensternberg-authentik reviewed (approved):
Well, the front-end changes look unremarkable. :-)
